lightning/src/ln/channelmanager.rs

```diff
@@ -3329,22 +3359,22 @@ impl<Signer: Sign, M: Deref, T: Deref, K: Deref, F: Deref, L: Deref> ChannelMana
 						// TODO: If we decided to blame ourselves (or one of our channels) in
 						// process_onion_failure we should close that channel as it implies our
 						// next-hop is needlessly blaming us!
-						self.pending_events.lock().unwrap().push(
-							events::Event::PaymentPathFailed {
-								payment_id: Some(payment_id),
-								payment_hash: payment_hash.clone(),
-								rejected_by_dest: !payment_retryable,
-								network_update,
-								all_paths_failed,
-								path: path.clone(),
-								short_channel_id,
-								retry,
+						let mut pending_events = self.pending_events.lock().unwrap();
+						pending_events.push(events::Event::PaymentPathFailed {
+							payment_id: Some(payment_id),
+							payment_hash: payment_hash.clone(),
+							rejected_by_dest: !payment_retryable,
+							network_update,
+							all_paths_failed,
+							path: path.clone(),
+							short_channel_id,
+							retry,
 #[cfg(test)]
-								error_code: onion_error_code,
+							error_code: onion_error_code,
 #[cfg(test)]
-								error_data: onion_error_data
-							}
-						);
+							error_data: onion_error_data
+						});
+						if let Some(ev) = full_failure_ev { pending_events.push(ev); }
```

Comment:
Could move this after the match to avoid duplicating it across patterns.

Reply:
I wanted to avoid holding the `pending_events` lock through the `process_onion_failure` call (cause it does crypto). If you feel strongly I can just eat it, but it feels wrong to do so.

Reply:
Gotcha. Alternatively, the match could evaluate to the `PaymentPathFailed` event and the lock could be held outside the match, with both events added then.